File: iMMPI/iMMPI/TestAnswersTableViewControllerBase.m
//
//  TestAnswersTableViewControllerBase.m
//  iMMPI
//
//  Created by Egor Chiglintsev on 04.01.13.
//  Copyright (c) 2013 Egor Chiglintsev. All rights reserved.
//

#if (!__has_feature(objc_arc))
#error "This file should be compiled with ARC support"
#endif

#import "TestAnswersTableViewControllerBase.h"


#pragma mark -
#pragma mark TestAnswersTableViewController implementation

@implementation TestAnswersTableViewControllerBase

#pragma mark -
#pragma mark methods

- (void) loadQuestionnaireAsyncIfNeeded: (dispatch_block_t) completion
{
    if ((_questionnaire == nil) && (_record != nil))
    {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            self.questionnaire = [Questionnaire newForGender: self.record.person.gender
                                                    ageGroup: self.record.person.ageGroup];
            
            if (completion)
                dispatch_async(dispatch_get_main_queue(), completion);
        });
    }
}


- (void) loadQuestionnaireAsyncIfNeeded
{
    [self loadQuestionnaireAsyncIfNeeded:^{
        [self.tableView reloadData];
    }];
}


- (id<StatementProtocol>) statementAtIndexPath: (NSIndexPath *) indexPath
{
    if ((indexPath.section == 0) && (indexPath.row < (NSInteger)self.questionnaire.statementsCount))
        return [_questionnaire statementAtIndex: indexPath.row];
    else
        return nil;
}


- (BOOL) saveRecord
{
    if (_record != nil)
        return [_storage updateTestRecord: _record];
    else
        return NO;
}


#pragma mark -
#pragma mark UITableViewDataSource

- (NSInteger) tableView: (UITableView *) tableView
  numberOfRowsInSection: (NSInteger) section
{
    return [_questionnaire statementsCount];
}


- (UITableViewCell *) tableView: (UITableView *) tableView
          cellForRowAtIndexPath: (NSIndexPath *) indexPath
{
    StatementTableViewCell *cell = (id)[tableView dequeueReusableCellWithIdentifier:
                                        [StatementTableViewCell reuseIdentifier]];
    FRB_AssertClass(cell, StatementTableViewCell);
    
    id<StatementProtocol> statement = [self statementAtIndexPath: indexPath];
    FRB_AssertNotNil(statement);
    
    cell.delegate                = self;
    cell.statementIDLabel.text   = [NSString stringWithFormat: @"%ld", (long)statement.statementID];
    cell.statementTextLabel.text = statement.text;
    
    switch ([_record.testAnswers answerTypeForStatementID: statement.statementID])
    {
        case AnswerTypePositive:
        {
            cell.statementSegmentedControl.selectedSegmentIndex = 1;
            cell.statementAnswerLabel.text = ___YES;
        } break;
            
            
        case AnswerTypeNegative:
        {
            cell.statementSegmentedControl.selectedSegmentIndex = 0;
            cell.statementAnswerLabel.text = ___NO;
        } break;
            
        case AnswerTypeUnknown:
        {
            cell.statementSegmentedControl.selectedSegmentIndex = UISegmentedControlNoSegment;
            cell.statementAnswerLabel.text = nil;
        } break;
    }

    return cell;
}


#pragma mark -
#pragma mark StatementTableViewCellDelegate

- (void) statementTableViewCell: (StatementTableViewCell *) cell
        segmentedControlChanged: (NSInteger) selectedSegmentIndex
{
    NSIndexPath *indexPath = [self.tableView indexPathForCell: cell];
    
    if (indexPath)
    {
        id<StatementProtocol> statement = [self statementAtIndexPath: indexPath];
        FRB_AssertNotNil(statement);
        
        switch (selectedSegmentIndex)
        {
            case 0: [_record.testAnswers setAnswerType: AnswerTypeNegative
                                        forStatementID: statement.statementID]; break;
                
            case 1: [_record.testAnswers setAnswerType: AnswerTypePositive
                                        forStatementID: statement.statementID]; break;
                
            default: [_record.testAnswers setAnswerType: AnswerTypeUnknown
                                         forStatementID: statement.statementID]; break;
        }
    }
}


#pragma mark -
#pragma mark SegueDestinationEditAnswers

- (void) setRecordToEditAnswers:  (id<TestRecordProtocol>) record
{
    self.record = record;
}


- (void) setStorageToEditAnswers: (id<TestRecordStorage>) storage
{
    self.storage = storage;
}


#pragma mark -
#pragma mark SegueSourceAnalyzeRecord

- (id<TestRecordProtocol>) recordForAnalysisWithSender: (id) sender
{
    return self.record;
}


- (id<TestRecordStorage>) storageForAnalysisWithSender: (id) sender
{
    return self.storage;
}

@end

File: iMMPI/iMMPI/TestAnswersTableViewControllerBase.swift
import UIKit

/// A base class for presenting and editing TestRecordProtocol record answers in a UITableView-driven UI.
///
/// This class uses StatementTableViewCell cells to display the statements of the questionnaire. 
/// The UITableViewDataSource methods are implemented to display the questionnaire contents in a single section list.
///
/// `TestAnswersTableViewControllerBase` is automatically set as the delegate for each of the `StatementTableViewCell`
/// used in the table view.
///
/// It is expected that the table view does return a `StatementTableViewCell` object for 
/// `StatementTableViewCell.reuseIdentifier()` string (this is set up in the storyboard).
///
/// **See also:** `TestAnswersInputViewController`, `TestAnswersViewController`.
class TestAnswersTableViewControllerBase: UIViewController, UsingRouting {
    weak var inputDelegate: TestAnswersInputDelegate?

    @IBOutlet var tableView: UITableView?

    var viewModel: TestAnswersViewModel? {
        willSet {
            viewModel?.onDidUpdate = Constant.void()
        }

        didSet {
            if let viewModel = viewModel {
                answers = viewModel.record.testAnswers.makeCopy()
                viewModel.onDidUpdate = { [weak self] in
                    self?.tableView?.reloadData()
                }
            }
        }
    }

    fileprivate(set) var answers = TestAnswers()
}


extension TestAnswersTableViewControllerBase {
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        viewModel?.setNeedsUpdate()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        saveRecord()
    }
}


extension TestAnswersTableViewControllerBase {
    func setAnswer(_ answer: AnswerType, for statement: Statement) {
        guard let record = viewModel?.record else {
            return
        }

        answers.setAnswer(answer, for: statement.statementID)
        inputDelegate?.testAnswersViewController(self, didSet: answer, for: statement, record: record)
    }
}


extension TestAnswersTableViewControllerBase {
    fileprivate func saveRecord() {
        guard let record = viewModel?.record else {
            return
        }

        inputDelegate?.testAnswersInputViewController(self, didSet: answers, for: record)
    }
}


// MARK: - UITableViewDataSource
extension TestAnswersTableViewControllerBase: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return viewModel?.statementsCount ?? 0
    }


    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: StatementTableViewCell.reuseIdentifier())
            as! StatementTableViewCell

        cell.delegate = self

        if let statement = viewModel?.statement(at: indexPath.row) {
            cell.statementIDLabel?.text = "\(statement.statementID)"
            cell.statementTextLabel?.text = statement.text

            switch answers.answer(for: statement.statementID) {
            case .positive:
                cell.statementAnswerLabel?.text = Strings.yes
                cell.statementSegmentedControl?.selectedSegmentIndex = 1

            case .negative:
                cell.statementAnswerLabel?.text = Strings.no
                cell.statementSegmentedControl?.selectedSegmentIndex = 0

            case .unknown:
                cell.statementAnswerLabel?.text = ""
                cell.statementSegmentedControl?.selectedSegmentIndex = UISegmentedControlNoSegment
            }
        }

        return cell
    }
}


extension TestAnswersTableViewControllerBase: StatementTableViewCellDelegate {
    func statementTableViewCell(_ cell: StatementTableViewCell, segmentedControlChanged selectedSegmentIndex: Int) {
        if let indexPath = tableView?.indexPath(for: cell), let statement = viewModel?.statement(at: indexPath.row) {
            switch selectedSegmentIndex {
            case 0: setAnswer(.negative, for: statement)
            case 1: setAnswer(.positive, for: statement)
            default: setAnswer(.unknown, for: statement)
            }
        }
    }
}
